// hookrouter/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "A", {
  enumerable: true,
  get: function get() {
    return _Link.A;
  }
});
Object.defineProperty(exports, "setLinkProps", {
  enumerable: true,
  get: function get() {
    return _Link.setLinkProps;
  }
});
Object.defineProperty(exports, "useRedirect", {
  enumerable: true,
  get: function get() {
    return _redirect.default;
  }
});
Object.defineProperty(exports, "useQueryParams", {
  enumerable: true,
  get: function get() {
    return _queryParams.useQueryParams;
  }
});
Object.defineProperty(exports, "setQueryParams", {
  enumerable: true,
  get: function get() {
    return _queryParams.setQueryParams;
  }
});
Object.defineProperty(exports, "getQueryParams", {
  enumerable: true,
  get: function get() {
    return _queryParams.getQueryParams;
  }
});
Object.defineProperty(exports, "useInterceptor", {
  enumerable: true,
  get: function get() {
    return _interceptor.useInterceptor;
  }
});
Object.defineProperty(exports, "useControlledInterceptor", {
  enumerable: true,
  get: function get() {
    return _controlledInterceptor.useControlledInterceptor;
  }
});
Object.defineProperty(exports, "useTitle", {
  enumerable: true,
  get: function get() {
    return _title.useTitle;
  }
});
Object.defineProperty(exports, "getTitle", {
  enumerable: true,
  get: function get() {
    return _title.getTitle;
  }
});
Object.defineProperty(exports, "navigate", {
  enumerable: true,
  get: function get() {
    return _router.navigate;
  }
});
Object.defineProperty(exports, "useRoutes", {
  enumerable: true,
  get: function get() {
    return _router.useRoutes;
  }
});
Object.defineProperty(exports, "setPath", {
  enumerable: true,
  get: function get() {
    return _router.setPath;
  }
});
Object.defineProperty(exports, "getPath", {
  enumerable: true,
  get: function get() {
    return _router.getPath;
  }
});
Object.defineProperty(exports, "getWorkingPath", {
  enumerable: true,
  get: function get() {
    return _router.getWorkingPath;
  }
});
Object.defineProperty(exports, "setBasepath", {
  enumerable: true,
  get: function get() {
    return _router.setBasepath;
  }
});
Object.defineProperty(exports, "getBasepath", {
  enumerable: true,
  get: function get() {
    return _router.getBasepath;
  }
});
Object.defineProperty(exports, "usePath", {
  enumerable: true,
  get: function get() {
    return _router.usePath;
  }
});

var _Link = require("./Link");

var _redirect = _interopRequireDefault(require("./redirect"));

var _queryParams = require("./queryParams");

var _interceptor = require("./interceptor");

var _controlledInterceptor = require("./controlledInterceptor");

var _title = require("./title");

var _router = require("./router");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
