// hookrouter/dist/interceptor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useInterceptor = exports.interceptRoute = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var incrementalId = 1;
var interceptors = [];

var interceptRoute = function interceptRoute(previousRoute, nextRoute) {
  if (!interceptors.length) {
    return nextRoute;
  }

  return interceptors.reduceRight(function (nextRoute, interceptor) {
    return nextRoute === previousRoute ? nextRoute : interceptor.handlerFunction(previousRoute, nextRoute);
  }, nextRoute);
};

exports.interceptRoute = interceptRoute;

var get = function get(componentId) {
  return interceptors.find(function (obj) {
    return obj.componentId === componentId;
  }) || null;
};

var remove = function remove(componentId) {
  var index = interceptors.findIndex(function (obj) {
    return obj.componentId === componentId;
  });

  if (index !== -1) {
    interceptors.splice(index, 1);
  }
};

var useInterceptor = function useInterceptor(handlerFunction) {
  var _React$useState = _react.default.useState(incrementalId++),
      _React$useState2 = _slicedToArray(_React$useState, 1),
      componentId = _React$useState2[0];

  var obj = get(componentId);

  if (!obj) {
    obj = {
      componentId: componentId,
      stop: function stop() {
        return remove(componentId);
      },
      handlerFunction: handlerFunction
    };
    interceptors.unshift(obj);
  }

  _react.default.useEffect(function () {
    return function () {
      return obj.stop();
    };
  }, []);

  return obj.stop;
};

exports.useInterceptor = useInterceptor;
