// hookrouter/dist/router.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useRoutes = exports.getWorkingPath = exports.usePath = exports.getPath = exports.setPath = exports.navigate = exports.ParentContext = exports.getBasepath = exports.setBasepath = void 0;

var _react = _interopRequireDefault(require("react"));

var _isNode = _interopRequireDefault(require("./isNode"));

var _queryParams = require("./queryParams");

var _interceptor = require("./interceptor");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var preparedRoutes = {};
var stack = {};
var componentId = 1;
var currentPath = _isNode.default ? '' : location.pathname;
var basePath = '';
var basePathRegEx = null;
var pathUpdaters = [];
/**
 * Will define a base path that will be utilized in your routing and navigation.
 * To be called _before_ any routing or navigation happens.
 * @param {string} inBasepath
 */

var setBasepath = function setBasepath(inBasepath) {
  basePath = inBasepath;
  basePathRegEx = new RegExp('^' + basePath);
};
/**
 * Returns the currently used base path.
 * @returns {string}
 */


exports.setBasepath = setBasepath;

var getBasepath = function getBasepath() {
  return basePath;
};

exports.getBasepath = getBasepath;

var resolvePath = function resolvePath(inPath) {
  if (_isNode.default) {
    var url = require('url');

    return url.resolve(currentPath, inPath);
  }

  var current = new URL(currentPath, location.href);
  var resolved = new URL(inPath, current);
  return resolved.pathname;
};

var ParentContext = _react.default.createContext(null);
/**
 * Pass a route string to this function to receive a regular expression.
 * The transformation will be cached and if you pass the same route a second
 * time, the cached regex will be returned.
 * @param {string} inRoute
 * @returns {Array} [RegExp, propList]
 */


exports.ParentContext = ParentContext;

var prepareRoute = function prepareRoute(inRoute) {
  if (preparedRoutes[inRoute]) {
    return preparedRoutes[inRoute];
  }

  var preparedRoute = [new RegExp("".concat(inRoute.substr(0, 1) === '*' ? '' : '^').concat(inRoute.replace(/:[a-zA-Z]+/g, '([^/]+)').replace(/\*/g, '')).concat(inRoute.substr(-1) === '*' ? '' : '$'))];
  var propList = inRoute.match(/:[a-zA-Z]+/g);
  preparedRoute.push(propList ? propList.map(function (paramName) {
    return paramName.substr(1);
  }) : []);
  preparedRoutes[inRoute] = preparedRoute;
  return preparedRoute;
};
/**
 * Virtually navigates the browser to the given URL and re-processes all routers.
 * @param {string} url The URL to navigate to. Do not mix adding GET params here and using the `getParams` argument.
 * @param {boolean} [replace=false] Should the navigation be done with a history replace to prevent back navigation by the user
 * @param {object} [queryParams] Key/Value pairs to convert into get parameters to be appended to the URL.
 * @param {boolean} [replaceQueryParams=true] Should existing query parameters be carried over, or dropped (replaced)?
 */


var navigate = function navigate(url) {
  var replace = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  var queryParams = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  var replaceQueryParams = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
  url = (0, _interceptor.interceptRoute)(currentPath, resolvePath(url));

  if (!url || url === currentPath) {
    return;
  }

  currentPath = url;

  if (_isNode.default) {
    setPath(url);
    processStack();
    updatePathHooks();
    return;
  }

  var finalURL = basePathRegEx ? url.match(basePathRegEx) ? url : basePath + url : url;
  window.history["".concat(replace ? 'replace' : 'push', "State")](null, null, finalURL);
  processStack();
  updatePathHooks();

  if (queryParams) {
    (0, _queryParams.setQueryParams)(queryParams, replaceQueryParams);
  }
};

exports.navigate = navigate;
var customPath = '/';
/**
 * Enables you to manually set the path from outside in a nodeJS environment, where window.history is not available.
 * @param {string} inPath
 */

var setPath = function setPath(inPath) {
  var url = require('url');

  customPath = url.resolve(customPath, inPath);
};
/**
 * Returns the current path of the router.
 * @returns {string}
 */


exports.setPath = setPath;

var getPath = function getPath() {
  return customPath;
};
/**
 * This hook returns the currently used URI.
 * Works in a browser context as well as for SSR.
 *
 * _Heads up:_ This will make your component render on every navigation unless you set this hook to passive!
 * @param {boolean} [active=true] Will update the component upon path changes. Set to false to only retrieve the path, once.
 * @param {boolean} [withBasepath=false] Should the base path be left at the beginning of the URI?
 * @returns {string}
 */


exports.getPath = getPath;

var usePath = function usePath() {
  var active = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
  var withBasepath = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

  var _React$useState = _react.default.useState(0),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      setUpdate = _React$useState2[1];

  _react.default.useEffect(function () {
    if (!active) {
      return;
    }

    pathUpdaters.push(setUpdate);
    return function () {
      var index = pathUpdaters.indexOf(setUpdate);

      if (index !== -1) {
        pathUpdaters.splice(index, 1);
      }
    };
  }, [setUpdate]);

  return withBasepath ? currentPath : currentPath.replace(basePathRegEx, '');
};
/**
 * Render all components that use path hooks.
 */


exports.usePath = usePath;

var updatePathHooks = function updatePathHooks() {
  var now = Date.now();
  pathUpdaters.forEach(function (cb) {
    return cb(now);
  });
};
/**
 * Called from within the router. This returns either the current windows url path
 * or a already reduced path, if a parent router has already matched with a finishing
 * wildcard before.
 * @param {string} [parentRouterId]
 * @returns {string}
 */


var getWorkingPath = function getWorkingPath(parentRouterId) {
  if (!parentRouterId) {
    return _isNode.default ? customPath : window.location.pathname.replace(basePathRegEx, '') || '/';
  }

  var stackEntry = stack[parentRouterId];

  if (!stackEntry) {
    throw 'wth';
  }

  return stackEntry.reducedPath !== null ? stackEntry.reducedPath || '/' : window.location.pathname;
};

exports.getWorkingPath = getWorkingPath;

var processStack = function processStack() {
  return Object.values(stack).forEach(process);
};
/**
 * This function takes two objects and compares if they have the same
 * keys and their keys have the same values assigned, so the objects are
 * basically the same.
 * @param {object} objA
 * @param {object} objB
 * @return {boolean}
 */


var objectsEqual = function objectsEqual(objA, objB) {
  var objAKeys = Object.keys(objA);
  var objBKeys = Object.keys(objB);

  var valueIsEqual = function valueIsEqual(key) {
    return objB.hasOwnProperty(key) && objA[key] === objB[key];
  };

  return objAKeys.length === objBKeys.length && objAKeys.every(valueIsEqual);
};

if (!_isNode.default) {
  window.addEventListener('popstate', function (e) {
    var nextPath = (0, _interceptor.interceptRoute)(currentPath, location.pathname);

    if (!nextPath || nextPath === currentPath) {
      e.preventDefault();
      e.stopPropagation();
      history.pushState(null, null, currentPath);
      return;
    }

    currentPath = nextPath;

    if (nextPath !== location.pathname) {
      history.replaceState(null, null, nextPath);
    }

    processStack();
    updatePathHooks();
  });
}

var emptyFunc = function emptyFunc() {
  return null;
};
/**
 * This will calculate the match of a given router.
 * @param {object} stackObj
 * @param {boolean} [directCall] If its not a direct call, the process function might trigger a component render.
 */


var process = function process(stackObj, directCall) {
  var routerId = stackObj.routerId,
      parentRouterId = stackObj.parentRouterId,
      routes = stackObj.routes,
      setUpdate = stackObj.setUpdate,
      resultFunc = stackObj.resultFunc,
      resultProps = stackObj.resultProps,
      previousReducedPath = stackObj.reducedPath;
  var currentPath = getWorkingPath(parentRouterId);
  var route = null;
  var targetFunction = null;
  var targetProps = null;
  var reducedPath = null;
  var anyMatched = false;

  for (var i = 0; i < routes.length; i++) {
    var _routes$i = _slicedToArray(routes[i], 2);

    route = _routes$i[0];
    targetFunction = _routes$i[1];

    var _ref = preparedRoutes[route] ? preparedRoutes[route] : prepareRoute(route),
        _ref2 = _slicedToArray(_ref, 2),
        regex = _ref2[0],
        groupNames = _ref2[1];

    var _result = currentPath.match(regex);

    if (!_result) {
      targetFunction = emptyFunc;
      continue;
    }

    if (groupNames.length) {
      targetProps = {};

      for (var j = 0; j < groupNames.length; j++) {
        targetProps[groupNames[j]] = _result[j + 1];
      }
    }

    reducedPath = currentPath.replace(_result[0], '');
    anyMatched = true;
    break;
  }

  if (!stack[routerId]) {
    return;
  }

  if (!anyMatched) {
    route = null;
    targetFunction = null;
    targetProps = null;
    reducedPath = null;
  }

  var funcsDiffer = resultFunc !== targetFunction;
  var pathDiffer = reducedPath !== previousReducedPath;
  var propsDiffer = true;

  if (!funcsDiffer) {
    if (!resultProps && !targetProps) {
      propsDiffer = false;
    } else {
      propsDiffer = !(resultProps && targetProps && objectsEqual(resultProps, targetProps) === true);
    }

    if (!propsDiffer) {
      if (!pathDiffer) {
        return;
      }
    }
  }

  var result = funcsDiffer || propsDiffer ? targetFunction ? targetFunction(targetProps) : null : stackObj.result;
  Object.assign(stack[routerId], {
    result: result,
    reducedPath: reducedPath,
    matchedRoute: route,
    passContext: route ? route.substr(-1) === '*' : false
  });

  if (!directCall && (funcsDiffer || propsDiffer || route === null)) {
    setUpdate(Date.now());
  }
};
/**
 * If a route returns a function, instead of a react element, we need to wrap this function
 * to eventually wrap a context object around its result.
 * @param RouteContext
 * @param originalResult
 * @returns {function(): *}
 */


var wrapperFunction = function wrapperFunction(RouteContext, originalResult) {
  return function () {
    return _react.default.createElement(RouteContext, null, originalResult.apply(originalResult, arguments));
  };
};
/**
 * Pass an object to this function where the keys are routes and the values
 * are functions to be executed when a route matches. Whatever your function returns
 * will be returned from the hook as well into your react component. Ideally you would
 * return components to be rendered when certain routes match, but you are not limited
 * to that.
 * @param {object} routeObj {"/someRoute": () => <Example />}
 */


var useRoutes = function useRoutes(routeObj) {
  // Each router gets an internal id to look them up again.
  var _React$useState3 = _react.default.useState(componentId),
      _React$useState4 = _slicedToArray(_React$useState3, 1),
      routerId = _React$useState4[0];

  var setUpdate = _react.default.useState(0)[1]; // Needed to create nested routers which use only a subset of the URL.


  var parentRouterId = _react.default.useContext(ParentContext); // If we just took the last ID, increase it for the next hook.


  if (routerId === componentId) {
    componentId += 1;
  } // Removes the router from the stack after component unmount - it won't be processed anymore.


  _react.default.useEffect(function () {
    return function () {
      return delete stack[routerId];
    };
  }, [routerId]);

  var stackObj = stack[routerId];

  if (stackObj && stackObj.originalRouteObj !== routeObj) {
    stackObj = null;
  }

  if (!stackObj) {
    stackObj = {
      routerId: routerId,
      originalRouteObj: routeObj,
      routes: Object.entries(routeObj),
      setUpdate: setUpdate,
      parentRouterId: parentRouterId,
      matchedRoute: null,
      reducedPath: null,
      passContext: false,
      result: null
    };
    stack[routerId] = stackObj;
    process(stackObj, true);
  }

  _react.default.useDebugValue(stackObj.matchedRoute);

  if (!stackObj.matchedRoute) {
    return null;
  }

  var result = stackObj.result;

  if (!stackObj.passContext) {
    return result;
  } else {
    var RouteContext = function RouteContext(_ref3) {
      var children = _ref3.children;
      return _react.default.createElement(ParentContext.Provider, {
        value: routerId
      }, children);
    };

    if (typeof result === 'function') {
      return wrapperFunction(RouteContext, result);
    }

    return _react.default.isValidElement(result) && result.type !== RouteContext ? _react.default.createElement(RouteContext, null, result) : result;
  }
};

exports.useRoutes = useRoutes;
